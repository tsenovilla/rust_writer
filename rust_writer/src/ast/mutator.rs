// SPDX-License-Identifier: GPL-3.0

use crate::Error;
use std::fmt::Debug;
use syn::{visit_mut::VisitMut, File};

#[derive(Debug, Clone)]
pub struct EmptyMutator;

pub struct Mutator<'a, T: Debug + Clone, const N: usize> {
	pub mutated: [bool; N],
	pub mutator: &'a T,
}

pub trait ToMutate<'a, T: Debug + Clone, const N: usize> {
	fn to_mutate(self, mutator: &'a T) -> Mutator<'a, T, N>;
}

impl Default for Mutator<'_, EmptyMutator, 1> {
	fn default() -> Self {
		Self { mutated: [false], mutator: &EmptyMutator }
	}
}

impl<'a, T, const N: usize> Mutator<'a, T, N>
where
	T: Debug + Clone,
	Mutator<'a, T, N>: VisitMut,
{
	pub fn mutate(mut self, ast: &mut File) -> Result<(), Error> {
		self.visit_file_mut(ast);

		if self.mutated.iter().all(|&x| x) {
			Ok(())
		} else {
			Err(Error::Descriptive(format!("Cannot mutate using Mutator: {:?}", self.mutator)))
		}
	}
}
