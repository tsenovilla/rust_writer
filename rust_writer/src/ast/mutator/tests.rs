// SPDX-License-Identifier: GPL-3.0

use super::*;
use crate::ast::implementors::ItemToTrait;
use syn::{parse_quote, TraitItem};
use test_builder::TestBuilder;

#[test]
fn mutator_works_for_a_implementor() {
	TestBuilder::default().with_trait_ast().execute(|mut builder| {
		let item_to_trait: ItemToTrait =
			("MyTrait", TraitItem::Type(parse_quote! {type Type1: From<String>;})).into();

		let ast = builder.get_mut_ast_file("trait.rs").expect("This exists; qed;");

		let mut mutator = Mutator::default().to_mutate(&item_to_trait);
		assert!(mutator.mutate(ast).is_ok());

		assert!(mutator.mutated[0]);

		mutator.reset();

		assert!(!mutator.mutated[0]);
	});
}
