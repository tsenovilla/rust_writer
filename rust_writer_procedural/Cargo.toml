[package]
name = "rust_writer_procedural"
version.workspace = true
repository.workspace = true 
license.workspace = true
rust-version.workspace = true
edition.workspace = true
description.workspace = true
documentation.workspace = true

[dependencies]
proc-macro2 = { workspace = true }
quote = { workspace = true }
syn = { workspace = true, features = ["full", "extra-traits", "parsing"] }

[lib]
proc-macro = true

[dev-dependencies]
trybuild = "1.0.103"
rust_writer = { workspace = true }
test_builder = { workspace = true }
