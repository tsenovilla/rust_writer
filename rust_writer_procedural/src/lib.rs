// SPDX-License-Identifier: GPL-3.0

mod finder;
pub(crate) mod helpers;
mod mutator;
pub(crate) mod parse;

use proc_macro::TokenStream;

/// # Description
///
/// The `#[mutator]` macro is used to define a new implementor which combines other implementors,
/// capable of mutating different parts of an AST with just one instruction, with the same level of
/// precision as if we had used each implementor separately.
///
/// The syntax to use the `#[mutator]` macro is pretty simple:
///
/// ```rust
/// use rust_writer::ast::{mutator, implementors::ItemToTrait, mutator::ToMutate};
/// use syn::visit_mut::VisitMut;
///
/// // The new implementor behaves as two `ItemToTrait` implementors.
/// #[mutator(ItemToTrait<'a>, ItemToTrait<'a>)]
/// struct NewImplementor;
/// ```
///
/// From the snippet above we can already assert that the `#[mutator]` macro needs that the traits
/// [`ToMutate`](https://docs.rs/rust_writer/latest/rust_writer/ast/mutator/trait.ToMutate.html)
/// and [`VisitMut`](https://docs.rs/syn/latest/syn/visit_mut/trait.VisitMut.html) need to be
/// brought into scope.
///
/// It's also possible to combine an implementor from the predefined set with a local implementor,
/// just by adding the keyword `local` before the local implementor.
///
/// ```rust
/// use rust_writer::ast::{mutator, local_mutator, implementors::ItemToTrait, mutator::ToMutate};
/// use syn::{visit_mut::VisitMut, File};
///
/// #[local_mutator]
/// #[derive(Debug, Clone)]
/// struct LocalImplementor<T: std::fmt::Debug + Clone>{
///   mutated: [bool; 4],
///   something: T
/// }
///
/// impl<T> VisitMut for LocalImplementor<T> where T: std::fmt::Debug + Clone{
///   fn visit_file_mut(&mut self, _file: &mut File){
///     // Dummy implementation
///     self.mutated = [true; 4];
///   }
/// }
///
/// // The new implementor behaves as two `ItemToTrait` implementors + our local implementor.
/// #[mutator(ItemToTrait<'a>, ItemToTrait<'a>, local = LocalImplementor<T: std::fmt::Debug + Clone>)]
/// struct NewImplementor;
/// ```
///
/// When the `#[mutator]` macro is called, the struct is expanded an includes a new field for each
/// implementor named as the implementor type in lowercase. If an implementor is included several
/// times in the same macro invocation, a suffix is added to the subsequent fields, such as
/// _1, _2, _3 and so on. For example, the snippet above struct will look like this after
/// the expansion:
///
/// ```ignore
/// struct NewImplementor<'a, T: std::fmt::Debug + Clone>{
///   itemtotrait: ItemToTrait<'a>,
///   itemtotrait_1: ItemToTrait<'a>,
///   localimplementor: LocalImplementor<T>
/// }
/// ```
///
/// Awesome, these snippets also show that if an implementor needs a generic, including it in the
/// macro invocation is enough. Particularly, the generics will be added to the `NewImplementor`
/// struct, so they'll be available inside it. It's also possible to place.
///
/// # The `#[impl_from]` attribute
///
/// # Compatibility with [`#[finder]`](https://docs.rs/rust_writer/latest/rust_writer/ast/attr.finder.html)
#[proc_macro_attribute]
pub fn mutator(attrs: TokenStream, item: TokenStream) -> TokenStream {
	mutator::mutator(attrs, item)
}

#[proc_macro_attribute]
pub fn local_mutator(_: TokenStream, item: TokenStream) -> TokenStream {
	mutator::local_mutator(item)
}

#[proc_macro_attribute]
pub fn finder(attrs: TokenStream, item: TokenStream) -> TokenStream {
	finder::finder(attrs, item)
}

#[proc_macro_attribute]
pub fn local_finder(attrs: TokenStream, item: TokenStream) -> TokenStream {
	finder::local_finder(attrs, item)
}

#[doc(hidden)]
#[proc_macro_attribute]
pub fn already_expanded(_: TokenStream, item: TokenStream) -> TokenStream {
	item
}
