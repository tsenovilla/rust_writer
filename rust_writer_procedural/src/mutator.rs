// SPDX-License-Identifier: GPL-3.0

mod expand;
mod parse;

use crate::parse_attrs::MacroAttrs;
use proc_macro::TokenStream;
use syn::{parse_macro_input, ItemStruct};

pub(crate) fn mutator(attrs: TokenStream, item: TokenStream) -> TokenStream {
	let attrs_list = parse_macro_input!(attrs as MacroAttrs);
	let struct_def = parse_macro_input!(item as ItemStruct);

	match parse::MutatorDef::try_from(attrs_list, struct_def) {
		Ok(def) => expand::expand_mutator(def).into(),
		Err(err) => err.to_compile_error().into(),
	}
}

pub(crate) fn impl_mutator(item: TokenStream) -> TokenStream {
	let mut finished = item.clone();

	let struct_def = parse_macro_input!(item as ItemStruct);

	let generated: TokenStream = match parse::ImplMutatorDef::try_from(struct_def) {
		Ok(def) => expand::expand_impl_mutator(def).into(),
		Err(err) => err.to_compile_error().into(),
	};

	finished.extend(generated);
	finished
}
