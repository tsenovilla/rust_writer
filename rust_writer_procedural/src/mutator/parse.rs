// SPDX-License-Identifier: GPL-3.0

use crate::parse_attrs::{InnerAttr, MacroAttrs};
use syn::{ItemStruct, Path, Result};

pub(crate) struct MutatorDef {
	pub(crate) implementors: Vec<Path>,
	pub(crate) struct_: ItemStruct,
	pub(crate) unit_struct: bool,
	pub(crate) impl_from: bool,
}

impl MutatorDef {
	pub(crate) fn try_from(attrs: MacroAttrs, mut struct_: ItemStruct) -> Result<Self> {
		let (unit_struct, impl_from) = match attrs.validate_struct(&mut struct_)? {
			InnerAttr::Unit => (true, false),
			InnerAttr::NotUnit => (false, false),
			InnerAttr::ImplFrom => (true, true),
			InnerAttr::ImplFromNotUnit => (false, true),
		};

		let implementors: Vec<Path> = attrs.0.into_iter().collect();

		Ok(Self { implementors, struct_, unit_struct, impl_from })
	}
}
