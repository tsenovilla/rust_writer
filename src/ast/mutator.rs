// SPDX-License-Identifier: GPL-3.0

use crate::Error;
use std::fmt::Debug;
use syn::{visit_mut::VisitMut, File};

#[derive(Debug)]
pub struct EmptyMutator;

pub struct Mutator<T: Debug> {
	pub mutated: bool,
	pub mutator: T,
}

pub trait ToMutate<T: Debug> {
	fn to_mutate(self, mutator: T) -> Mutator<T>;
}

impl Default for Mutator<EmptyMutator> {
	fn default() -> Self {
		Self { mutated: false, mutator: EmptyMutator }
	}
}

impl<T> Mutator<T>
where
	T: Debug,
	Mutator<T>: VisitMut,
{
	pub fn mutate(mut self, ast: &mut File) -> Result<(), Error> {
		self.visit_file_mut(ast);

		if self.mutated {
			Ok(())
		} else {
			Err(Error::Descriptive(format!("Cannot mutate using Mutator: {:?}", self.mutator)))
		}
	}
}
