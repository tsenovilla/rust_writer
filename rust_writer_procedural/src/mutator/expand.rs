// SPDX-License-Identifier: GPL-3.0

use crate::mutator::parse::MutatorDef;
use proc_macro2::{Span, TokenStream};
use quote::{quote, ToTokens};
use syn::{
	parse_quote, punctuated::Punctuated, token::Brace, Field, Fields, FieldsNamed, GenericParam,
	Ident, Index, Lifetime, PathArguments, Token, Type,
};

pub(crate) fn expand_mutator(def: MutatorDef) -> TokenStream {
	let MutatorDef { implementors, mut struct_, unit_struct, impl_from } = def;

	let struct_vis = &struct_.vis;
	let struct_name = &struct_.ident;

	let mutator_lifetime: Lifetime = parse_quote! {'mutator};
	let one = Index::from(1);

	let mut implementor_generics = Vec::new();

	let implementors_count = Index::from(implementors.iter().count());

	let implementors_indexes: Vec<Index> =
		implementors.iter().enumerate().map(|(index, _)| Index::from(index)).collect();

	let implementors_struct_names: Vec<Ident> = implementors
		.iter()
		.map(|implementor| {
			let last_implementor_segment = implementor
				.segments
				.last()
				.expect("At this point, implementors are valid paths; qed;");

			if let PathArguments::AngleBracketed(ref generics) = last_implementor_segment.arguments
			{
				generics.args.iter().for_each(|argument| {
					let generic_param: GenericParam = parse_quote!(#argument);
					if !implementor_generics.contains(&generic_param) {
						implementor_generics.push(generic_param);
					}
				})
			}

			Ident::new(
				&last_implementor_segment.ident.to_string().to_lowercase(),
				Span::call_site(),
			)
		})
		.collect();

	struct_.generics.params.extend(implementor_generics);

	let generics_params = &struct_.generics.params;

	let where_clause = struct_
		.generics
		.where_clause
		.as_ref()
		.map(|clause| clause.into_token_stream())
		.unwrap_or(quote! {where});

	let new_struct_fields: Vec<Field> = implementors_struct_names
		.iter()
		.zip(implementors)
		.map(|(name, implementor)| parse_quote!(#struct_vis #name: #implementor))
		.collect();

	if unit_struct {
		let mut fields: Punctuated<Field, Token![,]> = Punctuated::new();
		fields.extend(new_struct_fields);
		struct_.fields =
			Fields::Named(FieldsNamed { brace_token: Brace::default(), named: fields });
	} else {
		if let Fields::Named(ref mut fields) = struct_.fields {
			fields.named.extend(new_struct_fields);
		}
	}

	let mut impl_from_block = quote! {};

	if impl_from {
		if let Fields::Named(FieldsNamed { ref named, .. }) = struct_.fields {
			let fields_names: Vec<&Ident> = named
				.iter()
				.map(|field| field.ident.as_ref().expect("Named field has ident; qed;"))
				.collect();

			let fields_types: Punctuated<&Type, Token![,]> =
				named.iter().map(|field| &field.ty).collect();
			let tuple_indexes: Vec<Index> =
				named.iter().enumerate().map(|(index, _)| Index::from(index)).collect();

			impl_from_block = quote! {
				impl<#generics_params> From<(#fields_types)> for #struct_name<#generics_params> #where_clause{
					fn from(tuple: (#fields_types)) -> Self{
						Self{ #(#fields_names: tuple.#tuple_indexes),* }
					}
				}
			}
		}
	}


	let impl_to_mutate = quote! {
		impl<#mutator_lifetime, #generics_params>
		rust_writer::ast::mutator::ToMutate<#mutator_lifetime, #struct_name<#generics_params>, #implementors_count>
		for rust_writer::ast::mutator::Mutator<'_, rust_writer::ast::mutator::EmptyMutator, #one>
	  #where_clause
	  {
			fn to_mutate(self, mutator: &#mutator_lifetime #struct_name<#generics_params>)
			->
			rust_writer::ast::mutator::Mutator<#mutator_lifetime, #struct_name<#generics_params>,#implementors_count> {
				rust_writer::ast::mutator::Mutator {
					mutated: [false; #implementors_count],
				  mutator
				}
			}
		}
	};

	let impl_visit_mut = quote! {
		impl<#mutator_lifetime, #generics_params>
		syn::visit_mut::VisitMut
		for rust_writer::ast::mutator::Mutator<#mutator_lifetime, #struct_name<#generics_params>, #implementors_count>
	  #where_clause
	  {
			fn visit_file_mut(&mut self, file: &mut syn::File){
			#(
				  let mut mutator = rust_writer::ast::mutator::Mutator::default()
					  .to_mutate(&self.mutator.#implementors_struct_names);
				  mutator.visit_file_mut(file);
				  self.mutated[#implementors_indexes] = mutator.mutated.iter().all(|&x| x);
			)*
		 }
		}
	};

	quote! {
		#[rust_writer_procedural::already_expanded]
		#[derive(Debug, Clone)]
		#struct_
		#impl_from_block
		#impl_to_mutate
	  #impl_visit_mut
	}
	.into()
}
