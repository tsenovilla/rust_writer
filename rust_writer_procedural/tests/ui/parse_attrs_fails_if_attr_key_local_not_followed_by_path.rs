// SPDX-License-Identifier: GPL-3.0

use rust_writer_procedural::mutator;

#[mutator(rust_writer::ast::implementors::ItemToTrait<'a>, local = "hello")]
struct SomeStruct;

fn main() {}
