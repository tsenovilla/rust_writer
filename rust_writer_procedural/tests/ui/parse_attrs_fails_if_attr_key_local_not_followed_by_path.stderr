error: expected identifier
 --> tests/ui/parse_attrs_fails_if_attr_key_local_not_followed_by_path.rs:5:68
  |
5 | #[mutator(rust_writer::ast::implementors::ItemToTrait<'a>, local = "hello")]
  |                                                                    ^^^^^^^
